internal/controllers/machineset/machineset_controller.go

```diff
@@ -969,16 +993,54 @@ func (r *Reconciler) getMachineNode(ctx context.Context, cluster *clusterv1.Clus
 
 func (r *Reconciler) reconcileUnhealthyMachines(ctx context.Context, cluster *clusterv1.Cluster, ms *clusterv1.MachineSet, filteredMachines []*clusterv1.Machine) (ctrl.Result, error) {
 	log := ctrl.LoggerFrom(ctx)
+
+	// Calculate how many in flight machines we should remediate.
+	// By default, we allow all machines to be remediated at the same time.
+	maxInFlight := len(filteredMachines)
+
+	// If the MachineSet is part of a MachineDeployment, only allow remediations if
+	// it's the desired revision.
+	if util.HasOwner(ms.GetOwnerReferences(), clusterv1.GroupVersion.String(), []string{"MachineDeployment"}) {
+		owner, err := r.getOwnerMachineDeployment(ctx, ms)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if owner.Annotations != nil && ms.Annotations != nil && owner.Annotations[clusterv1.RevisionAnnotation] != ms.Annotations[clusterv1.RevisionAnnotation] {
+			// MachineSet is part of a MachineDeployment but isn't the current revision, no remediations allowed.
+			return ctrl.Result{}, nil
+		}
+
+		if owner.Spec.Strategy != nil && owner.Spec.Strategy.Remediation != nil {
+			if owner.Spec.Strategy.Remediation.MaxInFlight != nil {
+				var err error
+				maxInFlight, err = intstr.GetScaledValueFromIntOrPercent(owner.Spec.Strategy.Remediation.MaxInFlight, int(ptr.Deref(owner.Spec.Replicas, 1)), true)
+				if err != nil {
+					return ctrl.Result{}, fmt.Errorf("failed to calculate maxInFlight to remediate machines: %v", err)
+				}
+			}
+		}
+	}
+
 	// List all unhealthy machines.
 	machinesToRemediate := make([]*clusterv1.Machine, 0, len(filteredMachines))
 	for _, m := range filteredMachines {
+		if maxInFlight <= 0 {
+			break
+		}
+
 		// filteredMachines contains machines in deleting status to calculate correct status.
 		// skip remediation for those in deleting status.
 		if !m.DeletionTimestamp.IsZero() {
+			if conditions.IsTrue(m, clusterv1.MachineOwnerRemediatedCondition) {
+				// Machine has been remediated by this controller and still in flight.
+				maxInFlight--
```

Comment:
it'd be good to include a test case that accounts for maxInFlight in TestMachineSetReconciler_reconcileUnhealthyMachines

Reply:
done